Process.Kill: don't throw for exited processes

tmds opened the pull request:
Implements https://github.com/dotnet/corefx/issues/35257

CC @danmosemsft @wtgodbe @krwq

tmds commented:
I haven't tried this on Windows, so CI may point out some issues there.

stephentoub commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:
```diff
@@ -57,13 +57,30 @@ public static Process Start(string fileName, string arguments, string userName,
+
+            // Don't do anything if we know the process has exited.
+            // This also protects us from killing another process with a recycled PID.
+            if (GetHasExited(refresh: false))
```
Why `refresh: false` rather than `refresh: true`?

tmds replied:
This is a small optimization. Just below we make a `kill` call that will actively check if the process still exists, so here we can just check our _internal state_ without making a system call.

stephentoub replied:
The "This also protects us..." part makes it sounds like it's more than just an optimization.

tmds replied:
It's optimized compared to calling 'HasExited'. We need to check exited here for pid recycling. `EnsureState(State.HaveNonExitedId);` was also doing `refresh: false`.

I'll improve the comment.

stephentoub commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:
```diff
@@ -57,13 +57,30 @@ public static Process Start(string fileName, string arguments, string userName,
+                // Don't throw if the process has exited.
+                if (error != Interop.Error.ESRCH)
+                {
+                    throw new Win32Exception();
```
You didn't want to keep the comment?

tmds replied:
Not an active decision. I'll add it back.

stephentoub commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Win32.cs:
```diff
@@ -400,11 +406,6 @@ private IEnumerable<Exception> KillTree(SafeProcessHandle handle)
-            {
-                // The process isn't in a valid state for termination (e.g. already dead), so ignore the exception
-                // but don't give up in case children can still be enumerated.
-            }
```
This exception is no longer possible in any circumstance?

tmds replied:
If it does, I think it means something is broken with the implementation.

stephentoub replied:
Ok.  Let's make sure we're confident in our tests here then :wink:

tmds replied:
I looked closer at the implementation. `Kill` will only throw `InvalidOperationException` if the process is not associated or not local. All processes here will be associated & local, so no `InvalidOperationExceptions` are expected.

stephentoub replied:
Thanks for checking!

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:
```diff
@@ -783,6 +783,61 @@ public void TestProcessRecycledPid()
             Assert.True(foundRecycled);
         }
 
+        [Theory]
```
This test is Unix specific, but you changed the Windows implementation is well.  Is there no corresponding test on the Windows side?

tmds replied:
The Unix implementation has separate code paths for direct children vs other processes. For Windows, it is treated the same.

tmds replied:
In particular, how we detect process exits.

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:
```diff
@@ -783,6 +783,61 @@ public void TestProcessRecycledPid()
+                        // process still exists, wait some time.
+                        await Task.Delay(100);
+                    }
+                }
```
Loops like this very frequently come back to bite us, as at some point this is almost certain to fail in CI, with the reaping taking more than 10s at some point.  There must be a more deterministic way to do this, no?

tmds replied:
The corefx implementation is similar to what we are doing here (without a timeout). We can increase the timeout if 10s is too small.

danmosemsft replied:
I use a rule of thumb of a timeout that's maybe 10 times larger than it ever takes on a dev machine (up to maybe 30 sec) is necessary to make it solid on a CI machine.

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:
```diff
@@ -783,6 +783,61 @@ public void TestProcessRecycledPid()
+                        new RemoteInvokeOptions { StartInfo = new ProcessStartInfo() { RedirectStandardOutput = true }
+                        }))
+                    {
+                        Console.WriteLine(sleepProcess.Process.Id);
```
Can we avoid writing to stdout like this?  We try to only do so in error conditions, as otherwise it clutters up test output.

stephentoub replied:
Ah, you're redirecting output.  Nevermind.

tmds replied:
This stdout is redirected and read by the test, it doesn't end up as test output.

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs:
```diff
@@ -783,6 +783,61 @@ public void TestProcessRecycledPid()
+                {
+                    nonChildPid = int.Parse(createNonChildProcess.Process.StandardOutput.ReadToEnd());
+                }
+                return Process.GetProcessById(nonChildPid);
```
What would happen to the test if things got backed up in CI such that 30s passed and the sleepProcess actually completed / exited?

tmds replied:
The test expects the `Process.GetProcessById` and `kill SIGKILL` to happen before the `sleepProcess` exits (30 seconds currently). Otherwise it will throw.

stephentoub replied:
I'm worried we'll see this come back as sporadic failures in CI.  Can we instead make the child process wait for the parent process to exit (e.g. have the parent process pass its pid to the child and then have the child do Process.GetProcessById(pid).WaitForExit()?  If the test works correctly, the parent process should terminate the child process, and if something goes wrong and the parent process is terminated, the child process should exit.  It could also be some other scheme, I just like to avoid these timeout based tests, as they frequently lead to flakiness.

tmds replied:
>  Process.GetProcessById(pid).WaitForExit()

I'll add that. I'll also change the for-loop to a `while(true)`: since we successfully delivered SIGKILL, we expect the child to terminate eventually.

tmds commented:
I've removed `!HasExited` checks before calls to `Kill`. They were all in the tests. Except the one in the `Ping.Unix`. There I also removed the catch clauses, meant for dealing with exceptions due to the process exiting when calling `Kill`.

CC @wfurt

danmosemsft commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:
```diff
@@ -57,13 +57,34 @@ public static Process Start(string fileName, string arguments, string userName,
+                {
+                    return;
+                }
+
```
`GetLastError()` doesn't reset errno, does it? If it does, the line below should be `Win32Exception(error)`

tmds replied:
It doesn't change `errno`. This is an `Interop.Error` enum value. We can't pass this to `Win32Exception`. That one accepts the raw `int` errno value, from `Marshal.GetLastWin32Error`.

stephentoub commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:
```diff
@@ -88,10 +88,26 @@ public static void LeaveDebugMode()
         public void Kill()
         {
-            using (SafeProcessHandle handle = GetProcessHandle(Interop.Advapi32.ProcessOptions.PROCESS_TERMINATE))
+            using (SafeProcessHandle handle = GetProcessHandle(Interop.Advapi32.ProcessOptions.PROCESS_TERMINATE | Interop.Advapi32.ProcessOptions.PROCESS_QUERY_LIMITED_INFORMATION, throwIfExited: false))
```
Are there situations where the caller may have the rights necessary to use PROCESS_TERMINATE but not to use PROCESS_QUERY_LIMITED_INFORMATION?  I'm wondering if this change might be making it so that Kill stops working in cases it previously worked.

krwq replied:
i think only opposite is true

tmds replied:
TERMINATE is more impactful than QUERY_LIMITED_INFORMATION, so from that point also it makes sense if you can TERMINATE, you are likely to be able to QUERY_LIMITED_INFORMATION.
(QUERY_LIMITED_INFORMATION is also what is used to implement `HasExited`.)
